libsql-sqlite3/src/vectordiskann.c

```diff
@@ -0,0 +1,345 @@
+  if( pBuffer == NULL ){
+    rc = SQLITE_NOMEM_BKPT;
+    goto out;
+  }
+
+  // open blob in the end so we don't need to close it in error case
+  rc = sqlite3_blob_open(pIndex->db, pIndex->zDb, pIndex->zShadow, "data", nRowid, isWritable, &pBlobSpot->pBlob);
+  rc = blobSpotConvertRc(pIndex, rc);
+  if( rc != SQLITE_OK ){
+    goto out;
+  }
+  pBlobSpot->nRowid = nRowid;
+  pBlobSpot->pBuffer = pBuffer;
+  pBlobSpot->nBufferSize = nBufferSize;
+  pBlobSpot->isWritable = isWritable;
+  pBlobSpot->isInitialized = 0;
+
+  *ppBlobSpot = pBlobSpot;
+  return SQLITE_OK;
+
+out:
+  if( pBlobSpot != NULL ){
+    sqlite3_free(pBlobSpot);
+  }
+  if( pBuffer != NULL ){
+    sqlite3_free(pBuffer);
+  }
+  return rc;
+}
+
+int blobSpotReload(const DiskAnnIndex *pIndex, BlobSpot *pBlobSpot, u64 nRowid, int nBufferSize) {
+  int rc;
+
+  DiskAnnTrace(("blob spot reload: rowid=%lld\n", nRowid));
+  assert( pBlobSpot != NULL && pBlobSpot->pBlob != NULL );
+  assert( pBlobSpot->nBufferSize == nBufferSize );
+
+  if( pBlobSpot->nRowid == nRowid && pBlobSpot->isInitialized ){
+    return SQLITE_OK;
+  }
+
+  if( pBlobSpot->nRowid != nRowid ){
+    rc = sqlite3_blob_reopen(pBlobSpot->pBlob, nRowid);
+    rc = blobSpotConvertRc(pIndex, rc);
+    if( rc != SQLITE_OK ){
+      return rc;
+    }
+  }
+  rc = sqlite3_blob_read(pBlobSpot->pBlob, pBlobSpot->pBuffer, nBufferSize, 0);
+  if( rc != SQLITE_OK ){
+    return rc;
+  }
+  pBlobSpot->isInitialized = 1;
+  return SQLITE_OK;
+}
+
+int blobSpotFlush(BlobSpot *pBlobSpot) {
+  return sqlite3_blob_write(pBlobSpot->pBlob, pBlobSpot->pBuffer, pBlobSpot->nBufferSize, 0);
+}
+
+void blobSpotFree(BlobSpot *pBlobSpot) {
+  if( pBlobSpot->pBlob != NULL ){
+    sqlite3_blob_close(pBlobSpot->pBlob);
+  }
+  if( pBlobSpot->pBuffer != NULL ){
+    sqlite3_free(pBlobSpot->pBuffer);
+  }
+  sqlite3_free(pBlobSpot);
+}
+
+/**************************************************************************
+** Layout specific utilities
+**************************************************************************/
+
+int nodeEdgesMaxCount(const DiskAnnIndex *pIndex){
+  unsigned int nMaxEdges = (pIndex->nBlockSize - pIndex->nNodeVectorSize - VECTOR_NODE_METADATA_SIZE) / (pIndex->nEdgeVectorSize + VECTOR_EDGE_METADATA_SIZE);
+  assert( nMaxEdges > 0);
+  return nMaxEdges;
+}
+
+int nodeEdgesMetadataOffset(const DiskAnnIndex *pIndex){
+  unsigned int offset;
+  unsigned int nMaxEdges = nodeEdgesMaxCount(pIndex);
+  offset = VECTOR_NODE_METADATA_SIZE + pIndex->nNodeVectorSize + nMaxEdges * pIndex->nEdgeVectorSize;
+  assert( offset <= pIndex->nBlockSize );
+  return offset;
+}
+
+void nodeBinInit(const DiskAnnIndex *pIndex, BlobSpot *pBlobSpot, u64 nRowid, Vector *pVector){
+  assert( VECTOR_NODE_METADATA_SIZE + pIndex->nNodeVectorSize <= pBlobSpot->nBufferSize );
+
+  memset(pBlobSpot->pBuffer, 0, pBlobSpot->nBufferSize);
+  writeLE64(pBlobSpot->pBuffer, nRowid);
+  writeLE16(pBlobSpot->pBuffer + sizeof(u64), 0);
```

Comment:
Isn't this already 0 after memset from line 223?

Reply:
Will remove this and leave comment that neighbours count already correctly set to zero